### PR TITLE
Cache queries

### PR DESCRIPTION
**Pull request by k0ka**
- [x] Added or updated tests
- [x] Documented user facing changes
- [x] Updated CHANGELOG.md

**Changes**

Query caching has been added: the result of `GraphQL\Language\Parser::parse()` will be serialized and stored in the Laravel cache. It can be disabled in the config.

I tested it on my production queries and it shows about 5-10% performance boost. I use Relay with a lot of fragments. 

I use sha256 hash of query as a key. So we may develop it further towards implementing [Apollo Automatic persisted queries](https://www.apollographql.com/docs/apollo-server/performance/apq/). While Relay currently uses md5 hashes for their [Persisted queries](https://relay.dev/docs/next/guides/persisted-queries/), I guess they will [change it](https://github.com/facebook/relay/pull/3666) soon to a better hashing algorithm. 

**Breaking changes**

`GraphQL::executeQuery()` is deprecated and should be removed in the next version. `parseAndExecuteQuery()` or `executeParsedQuery()` should be used instead. This PR has a simple `executeQuery()` stub that executes one of above methods depending on the `query` type.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,10 @@ You can find and compare releases at the [GitHub release page](https://github.co
 
 ## Unreleased
 
+### Added
+
+- Added the query caching
```
Comment (spawnia) [inline CHANGELOG.md:14]:
```suggestion
- Added query caching
```

File: UPGRADE.md [hunk 2]
```diff
@@ -126,6 +126,11 @@ the new behaviour before upgrading by setting `unbox_bensampo_enum_enum_instance
 public function scopeByType(Builder $builder, AOrB $aOrB): Builder
 ```
 
+### Nuwave\Lighthouse\GraphQL::executeQuery() is deprecated
+
+Use `parseAndExecuteQuery()` for executing a string query or `executeParsedQuery()` for 
+executing already parsed `DocumentNode`
```
Comment (spawnia) [inline UPGRADE.md:132]:
```suggestion
executing already parsed `DocumentNode`.
```

File: docs/master/performance/query-caching.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# Query caching
+
+In order to speed up the GraphQL query parsing, its results can be stored in the Laravel cache.
+
+Query caching is enabled in non-local environments by default. You can define cache store and cache duration,
+see `config/lighthouse.php`.
+
+The GraphQL schema doesn't affect query parsing, so you have to flush query cache only on the package update.
```
Comment (spawnia) [inline docs/master/performance/query-caching.md:8]:
```suggestion
Make sure to flush the query cache when you deploy an upgraded version of the dependency `webonyx/graphql-php`:

    some-one-liner that clears the cache?
```

File: src/GraphQL.php [hunk 4]
```diff
@@ -101,7 +103,7 @@ public function __construct(
     /**
      * Run one ore more GraphQL operations against the schema.
      *
-     * @param  \GraphQL\Server\OperationParams|array<int, \GraphQL\Server\OperationParams>  $operationOrOperations
```
Comment (spawnia) [inline src/GraphQL.php:104]: please revert to the previous formatting in `@param`

File: src/lighthouse.php [hunk 5]
```diff
@@ -126,6 +126,37 @@
...
+    |
+    | Caches user queries so we won't have to parse string queries on each request
+    |
+    */
+
+    'query_cache' => [
+        /*
+         * Setting to true enables query caching.
+         */
+        'enable' => env('LIGHTHOUSE_QUERY_CACHE_ENABLE', 'local' !== env('APP_ENV')),
```
Comment (spawnia) [inline src/lighthouse.php:142]: why not local?

File: src/lighthouse.php [hunk 6]
```diff
@@ -126,6 +126,37 @@
...
+        'store' => env('LIGHTHOUSE_QUERY_CACHE_STORE', null),
+
+        /*
+         * Duration in seconds the query should remain cached, null means forever.
+         */
+        'ttl' => env(
+            'LIGHTHOUSE_QUERY_CACHE_TTL',
+            \Nuwave\Lighthouse\Support\AppVersion::atLeast(5.8)
+            ? null
+            : 365 * 24 * 60 // For Laravel < 5.8 the exact value must be specified and it is counted in minutes
```
Comment (spawnia) [inline src/lighthouse.php:156]:
```suggestion
                ? null
                : 365 * 24 * 60 // For Laravel < 5.8 the exact value must be specified and it is counted in minutes
```

**Comment by spawnia** [comment]
I need to think about the deprecation of `executeQuery()` and the naming of the new methods.

Perhaps we should extract a method `parse(string $query): DocumentNode`?

File: docs/master/performance/query-caching.md (hunk 3 above)
Comment (k0ka) [reply docs/master/performance/query-caching.md]: I can add another command like `lighthouse:query:cache:clear` but it only clears the cache store. So I guess it's easier to just use the laravel default command.

File: src/lighthouse.php (hunk 5 above)
Comment (k0ka) [reply src/lighthouse.php:142]: I just copied it from schema caching. It works on any env and all tests are passed with enabled query cache.

**Comment by k0ka** [comment]
> Perhaps we should extract a method `parse(string $query): DocumentNode`?

This method might also throw a `SyntaxError` which must be caught and wrapped in the `ExecutionResult`. So it's better to leave `parseAndExecuteQuery()` and it is not too big and complex to split it.

File: src/lighthouse.php (hunk 5 above)
Comment (spawnia) [reply src/lighthouse.php:142]:
```suggestion
        'enable' => env('LIGHTHOUSE_QUERY_CACHE_ENABLE', true),
```

File: docs/master/performance/query-caching.md (hunk 3 above)
Comment (spawnia) [reply docs/master/performance/query-caching.md]: That Laravel command clears all caches?

**Comment by spawnia** [comment]
I could imagine the memoized `parse()` method being useful of its own.

File: docs/master/performance/query-caching.md (hunk 3 above)
Comment (k0ka) [reply docs/master/performance/query-caching.md]: it fully clears the default store. If the user specified non-default store for query cache they should add an argument to the command
`php artisan cache:clear --store=query_cache_store`

**Comment by k0ka** [comment]
I extracted the `GraphQL::parse()` method

File: src/GraphQL.php [hunk 7]
```diff
@@ -185,15 +188,77 @@ public function executeQuery(
...
+            $parsedQuery = $this->parse($query);
+        } catch (SyntaxError $syntaxError) {
+            return new ExecutionResult(null, [$syntaxError]);
         }
 
+        return $this->executeParsedQuery($parsedQuery, $context, $variables, $rootValue, $operationName);
+    }
+
+    /**
+     * Parses string query to DocumentNode. Caches queries if required.
```
Comment (spawnia) [inline src/GraphQL.php:221]:
```suggestion
     * Parse the given query string into a DocumentNode.
     *
     * Caches the parsed result if the query cache is enabled in the configuration.
```

File: src/GraphQL.php [hunk 8]
```diff
@@ -185,15 +188,77 @@ public function executeQuery(
...
+        if (! $cacheConfig['enable']) {
+            return Parser::parse($query);
+        }
+
+        /** @var \Illuminate\Contracts\Cache\Factory $cacheFactory */
+        $cacheFactory = app(CacheFactory::class);
+        $store = $cacheFactory->store($cacheConfig['store']);
+
+        return $store->remember(
+            'lighthouse:query:' . hash('sha256', $query),
```
Comment (spawnia) [inline src/GraphQL.php:238]: We don't use global helpers such as `hash()`, they are usually not compatible with Lumen.

File: src/GraphQL.php [hunk 9]
```diff
@@ -185,15 +188,77 @@ public function executeQuery(
...
+            $cacheConfig['ttl'],
+            function () use ($query) {
```
Comment (spawnia) [inline src/GraphQL.php:240]:
```suggestion
            static function () use ($query): DocumentNode {
```

File: src/lighthouse.php [hunk 10]
```diff
@@ -126,6 +126,37 @@
         'tags' => false,
     ],
 
+    /*
+    |--------------------------------------------------------------------------
+    | Query Cache
+    |--------------------------------------------------------------------------
+    |
+    | Caches user queries so we won't have to parse string queries on each request
```
Comment (spawnia) [inline src/lighthouse.php:134]:
```suggestion
    | Caches the result of parsing incoming query strings to boost performance on subsequent requests.
```

**Review by spawnia — COMMENTED** [review]
Like the extracted method, just a few more nits regarding the docs.

Considering [Automated persisted queries](https://github.com/nuwave/lighthouse/issues/651), would you foresee further changes to the proposed API?

File: src/GraphQL.php (hunk 8 above)
Comment (k0ka) [reply src/GraphQL.php:240]: `hash()` is a global php function - https://www.php.net/manual/en/function.hash.php

**Comment by k0ka** [comment]
> Considering [Automated persisted queries](https://github.com/nuwave/lighthouse/issues/651), would you foresee further changes to the proposed API?

I don't thinks this API will have any changes. We will have to add another method like `getParsedQueryByHash(string $sha256hash): DocumentNode`. It should get the corresponding cache value and return apollo compatible error if this value is not present in the cache.

**Comment by spawnia** [comment]
The API of the `GraphQL` class is one thing that might have to change, but I am also concerned about the configuration format. We have some questionable structure in the config file (`schema.register`...) already. I would like to implement both features at once to make sure they fit together nicely, The config for APQ would probably overlap with `query_cache` quite a bit.

I can look into this myself, but am also quite happy if you push it further.

File: src/GraphQL.php (hunk 8 above)
Comment (spawnia) [reply src/GraphQL.php:240]: You are right, thanks!

**Comment by k0ka** [comment]
I added the APQ support. `webonyx/graphql-php` already supports it so it was quite easy. 

As for config schema: I guess the better cache format will be like this:
```php
'cache' => [
    'schema' => [
          'version' => env('LIGHTHOUSE_CACHE_VERSION', 1),

          // other parameters that are currently in the cache.*
    ],

    'query' => [
          // parameters from query_cache
    ],

    'validate' => [
         // future parameters for https://github.com/nuwave/lighthouse/issues/2018
    ],
]
```

But it will break backward compatibility. So I guess it's easier to leave as is and refactor it in the next major version.

**Review by spawnia — APPROVED** [review]
Awesome job!

**Comment by spawnia** [comment]
@k0ka invited you as a collaborator. If you would like to [join our Slack](https://join.slack.com/t/lighthouse-php/shared_invite/zt-4sm280w1-wu21r94f3kLRRtBXRbXVfw), I can add you to our collaborator chat as well.

**Pull request merged**